internal/konnect/node_agent_test.go

```diff
@@ -0,0 +1,406 @@
+					Type:     NodeTypeIngressController,
+					Status:   string(IngressControllerStateOperational),
+					Version:  testKicVersion,
+				},
+			},
+			numNodes: 1,
+		},
+		{
+			name:     "update gateway nodes and remove outdated nodes",
+			hostname: "ingress-0",
+			initialNodes: []*NodeItem{
+				{
+					Hostname: "ingress-0",
+					ID:       uuid.NewString(),
+					Type:     NodeTypeIngressController,
+					Status:   string(IngressControllerStateOperational),
+					Version:  testKicVersion,
+				},
+				{
+					Hostname: "proxy-0",
+					ID:       uuid.NewString(),
+					Type:     NodeTypeKongProxy,
+					Version:  testKongVersion,
+				},
+				// 2 gateway nodes with same name, should reserve newer one.
+				{
+					Hostname: "proxy-1",
+					ID:       uuid.NewString(),
+					Type:     NodeTypeKongProxy,
+					Version:  testKongVersion,
+					LastPing: time.Now().Unix() - 10,
+				},
+				{
+					Hostname: "proxy-1",
+					ID:       uuid.NewString(),
+					Type:     NodeTypeKongProxy,
+					Version:  testKongVersion,
+					LastPing: time.Now().Unix() - 5,
+				},
+			},
+			gatewayInstances: []GatewayInstance{
+				{hostname: "proxy-1", version: testKongVersion},
+			},
+			containNodes: []*NodeItem{
+				{
+					Hostname: "ingress-0",
+					Type:     NodeTypeIngressController,
+					Status:   string(IngressControllerStateOperational),
+					Version:  testKicVersion,
+				},
+				{
+					Hostname: "proxy-1",
+					Type:     NodeTypeKongProxy,
+					Version:  testKongVersion,
+				},
+			},
+			notContainNodes: []*NodeItem{
+				{
+					Hostname: "proxy-0",
+					Type:     NodeTypeKongProxy,
+					Version:  testKongVersion,
+				},
+			},
+			numNodes: 2,
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			nodeService := &mockKonnectNodeService{
+				clusterID: testClusterID,
+				nodes:     tc.initialNodes,
+			}
+			s := httptest.NewServer(nodeService)
+			nodeClient := &NodeAPIClient{
+				Address:        s.URL,
+				RuntimeGroupID: testClusterID,
+				Client:         &http.Client{},
+			}
+
+			configStatusSubscriber := dataplane.NewChannelConfigNotifier()
+
+			nodeAgent := NewNodeAgent(
+				tc.hostname, testKicVersion,
+				DefaultRefreshNodePeriod,
+				testr.New(t),
+				nodeClient,
+				configStatusSubscriber,
+				&mockGatewayInstanceGetter{tc.gatewayInstances},
+			)
+			go func() {
+				err := nodeAgent.Start(context.Background())
+				require.NoError(t, err)
+			}()
+
+			if tc.configStatus != nil {
+				configStatusSubscriber.NotifyConfigStatus(*tc.configStatus)
+			}
+
+			require.Eventually(t, func() bool {
+				err := nodeAgent.updateNodes()
+				require.NoError(t, err)
+				// check number of nodes in RG.
+				if len(nodeService.nodes) != tc.numNodes {
+					return false
+				}
+				// check for nodes that must be included in RG by hostname, type, version and status.
+				for _, node := range tc.containNodes {
+					if !lo.ContainsBy(
+						nodeService.nodes,
```

Comment:
Shouldn't we make sure that access to `nodeService.nodes` is concurrency-safe? It can potentially cause race conditions because we modify the nodes in the server goroutine and read it in the test. I think we could add a getter to the `nodeService` that would return a copy of the nodes slice and guard access to the internal slice with a mutex. WDYT?

Reply:
added locks.